### PR TITLE
[autodiff] Support autodiff for torch Tensor and taichi ndarray on CPU and CUDA

### PR DESCRIPTION
**Pull request by ailzhang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #7889
* #7888
* #7887
* __->__ #7878

Known limitation:
- Only support torch Tensor & ti.ndarray on CPU & CUDA backends, SPIRV
  backend support will be in a separate PR.
-

### CODE REVIEW

**Review by erizmr — COMMENTED** [review]
Thanks!

File: python/taichi/lang/kernel_impl.py [hunk 1]
```diff
@@ -706,6 +707,10 @@ def call_back():
 
                                 return call_back
 
+                            # FIXME: only allocate when launching grad kernel
+                            if v.requires_grad and v.grad is None:
+                                v.grad = torch.zeros_like(v)
```
[the remark this answers is not in the document]
Comment (ailzhang) [reply python/taichi/lang/kernel_impl.py:712]: IIUC the simple cases should just work, added an example at https://github.com/taichi-dev/taichi/pull/7889/files ;)

File: python/taichi/ad/_ad.py [hunk 2]
```diff
@@ -186,22 +187,30 @@ def __enter__(self):
...
+                clear_all_gradients()
+            if self.validation:
+                clear_all_gradients(gradient_type=SNodeGradType.ADJOINT_CHECKBIT)
+
+            from taichi._kernels import clear_loss  # pylint: disable=C0415
 
-        from taichi._kernels import clear_loss  # pylint: disable=C0415
+            clear_loss(self.loss)
+        elif isinstance(self.loss, Ndarray):
+            self.loss.fill(0.0)
```
[the remark this answers is not in the document]
Comment (ailzhang) [reply python/taichi/ad/_ad.py:208]: SG! Added the check in https://github.com/taichi-dev/taichi/pull/7889/files

File: taichi/program/launch_context_builder.cpp [hunk 3]
```diff
@@ -77,7 +77,9 @@ void LaunchContextBuilder::set_ndarray_ptrs(int arg_id,
                                             uint64 data_ptr,
                                             uint64 grad_ptr) {
   set_struct_arg({arg_id, TypeFactory::DATA_PTR_POS_IN_NDARRAY}, data_ptr);
```
Comment (lin-hitonami) [inline taichi/program/launch_context_builder.cpp:79]: Should we assert data_ptr and grad_ptr != 0 here

**Review by lin-hitonami — APPROVED** [review]
LGTM (didn't review the autodiff part)

**Review by erizmr — APPROVED** [review]
LGTM for the Autodiff part.

File: taichi/program/launch_context_builder.cpp (hunk 3 above)
Comment (ailzhang) [reply taichi/program/launch_context_builder.cpp:79]: SG! Will add it in the next PR!

**Comment by taichi-gardener** [comment]
ghstack bot failed: PR #7878 check-run `Build and Test (AMDGPU)`'s status `None` is not success!

**Pull request merged**